api/desecapi/authentication.py

```diff
@@ -16,6 +17,20 @@
 class TokenAuthentication(RestFrameworkTokenAuthentication):
     model = Token
 
+    def authenticate(self, request):
+        try:
+            user, token = super().authenticate(request)
+        except TypeError:  # TypeError: cannot unpack non-iterable NoneType object
+            return None  # unauthenticated
+
+        client_ip = ip_address(request.META.get('REMOTE_ADDR'))
```

Comment:
It's indeed unfortunate that `request.META` contains items of different nature. It gave me a headache in some other circumstance (I don't remember what it was).

Anyways: Complications with client IPs usually come in when the remote address is that of an application proxy, which puts the real client IP into the `X-Forwarded-For` (or similar) header. When present, some applications use that as the client IP, ignoring the remote address. This is, obviously, potentially vulnerable to spoofing. In the context of rate limiting, we therefore took care to [properly configure things](https://github.com/desec-io/desec-stack/blob/cac3133/api/api/settings.py#L116), [as explained in the DRF docs](https://www.django-rest-framework.org/api-guide/throttling/#how-clients-are-identified).

This is not an issue in the above code because it uses the remote address itself. However, that also means that the functionality will break if the stack is run behind an application proxy. For rate limiting, it's just a change of a setting, whereas subnet-based token authentication will break.

Maybe we should consider using the mechanism used by DRF to determine the client IP address? It may be possible to call it without copying the code itself. Aside from the proxy use case, this would also give a higher degree of consistency. What do you think?

Reply:
Let's put an excerpt of this discussion along with important sources in the code.

I'd stick to `request.META['REMOTE_ADDR']` for now, as it errs on the safe side. We could add a hint to the README about the proxies. Or create/modify an api-specific README, warning that things will break if put behind a proxy.

Reply:
done